### ISSUE TITLE
Clarification: Does the 14-dim hand action go through the decoder ONNX during deployment vs sim/training? And what's needed to swap hands (e.g., Inspire Hand)?

### ISSUE THREAD

**Issue by Gumin-Research** [NVlabs/GR00T-WholeBodyControl#159]
## Background

I'm trying to understand the exact data flow for the 14-dim hand action (7 left + 7 right Dex3) across three scenarios: real-robot deployment, sim, and SONIC training. The answer affects what would need to change to support a different hand (e.g., Inspire Hand with a different DoF count, or Unitree Dex5).

### Scenario A. Real-robot deployment (C++)

In `gear_sonic_deploy/src/g1/g1_deploy_onnx_ref/src/g1_deploy_onnx_ref.cpp`, hand joints come from the input interface (ZMQ / keyboard / VR) and are sent **directly to the Dex3 driver**, bypassing the decoder ONNX entirely:

```cpp
// L2969-2970: read hand joints from input source
std::tie(has_left_hand_data_,  left_hand_joint_buffer_)  = input_interface_->GetHandPose(true);
std::tie(has_right_hand_data_, right_hand_joint_buffer_) = input_interface_->GetHandPose(false);

// L3954-3955: push them straight to Dex3
dex3_hands_.setAllJointsCommand(true,  left_hand_joint_buffer_);
dex3_hands_.setAllJointsCommand(false, right_hand_joint_buffer_);
```

The decoder ONNX only outputs `G1_NUM_MOTOR = 29` body actions (`g1_deploy_onnx_ref.cpp:3119-3128`).

Hand DoF is hard-coded to 7 per side:

```cpp
// gear_sonic_deploy/.../include/dex3_hands.hpp:51
static constexpr int DEX3_MOTOR_MAX = 7;
```

ZMQ Protocol v4 also fixes the hand shape to `[7]` per hand (`docs/tutorials/zmq.md:300-302`).

### Scenario B. Training / sim (Python)

Things look quite different on the training/sim side. In `gear_sonic/envs/wrapper/manager_env_wrapper.py:668-696`, the policy's `meta_actions` are split into **tokenizer (body) + hand parts**, and the hand part is processed inside the same forward pass:

```python
tokenizer_action_dim = self.config.get("tokenizer_action_dim")
tokenizer_meta_actions = meta_actions[:, :tokenizer_action_dim]
hand_actions_raw = meta_actions[:, tokenizer_action_dim:]

# Optionally override with motion data
if self.config.get("use_motion_hand_actions", False):
    left_action  = motion_cmd.get_hand_action("left_hand")
    right_action = motion_cmd.get_hand_action("right_hand")
    hand_actions_raw = torch.stack([left_action, right_action], dim=-1)

# Optional primitive to finger joint mapping
if self._use_finger_primitive and self._finger_primitive_map:
    hand_actions = self._convert_primitive_to_finger_actions(hand_actions_raw)
else:
    hand_actions = hand_actions_raw

# Later: concat with body
env_actions = torch.cat([body_actions, hand_actions], dim=-1)
```

There's also a `hand_dyn` decoder defined alongside `g1_dyn` in `gear_sonic/trl/modules/universal_token_modules.py:909-921` that concatenates its `hand_action` onto the body action during training. So at training time hands do go through a decoder (`hand_dyn`), and there's also a primitive to finger mapping layer.

### Scenario C. ONNX export

`gear_sonic/eval_agent_trl.py:505, 513, 521, 535`: all four export calls hardcode `decoder_name="g1_dyn"`. A repo-wide search for `decoder_name="hand_dyn"` returns nothing. So `hand_dyn` and the primitive mapping never make it into the deployed ONNX.

## Questions

1. **Deployment (real robot):** Is it correct that the 14-dim hand action **never** passes through any ONNX model at deploy time, and goes straight from the input interface (VLA / teleop / keyboard) to `dex3_hands_`?

2. **Sim:** In MuJoCo sim with `--input-type keyboard sim`, do hands behave the same way (bypass decoder), or does the sim path actually invoke `hand_dyn` / the primitive to finger mapping shown above?

3. **Training:** During SONIC RL training, hands clearly go through `hand_dyn` plus the optional primitive to finger mapping. Is `hand_dyn` purely an auxiliary signal to shape the encoder, or is its output ever used to supervise the VLA fine-tune (so that the VLA learns to emit hand joints directly at inference)?

4. **Swapping hands (e.g., Inspire Hand or Unitree Dex5):** assuming the above is right, the things that would need to change for a new hand are:

   - C++ driver: replace `dex3_hands.hpp` / update `DEX3_MOTOR_MAX`
   - Input interfaces (ZMQ / keyboard / VR): emit the new hand's DoF instead of `[7]`
   - `observation_config.yaml`: any hand-related fields
   - VLA dataset & fine-tune: collect new `left/right_hand_joints` with the new DoF, then refine the VLA
   - Training (only if going beyond VLA): retrain `hand_dyn` for the new DoF; possibly retrain the primitive to finger map

   **The decoder ONNX itself (`decoder.onnx`, body-only 29-dim) should NOT need retraining**, right? It's purely body. Just confirming.

5. **Is there an embodiment-portability / hand-swap guide?** I couldn't find one in the docs and want to plan correctly before any hardware change.

## Setup

- GR00T-WholeBodyControl `main` branch (cloned 2026-05-18)
- Embodiment: `UNITREE_G1_SONIC` (Unitree G1 + Dex3)
- Hardware: Jetson Orin NX 16GB, JetPack 6.2, TensorRT 10.7

Thanks for the great release!

**Comment by chenran-nvidia** [comment]
1. The hand action is passed directly from the interface to the robot; it is not controlled by the Sonic policy. If you are using the VLA setup, then the VLA controls it directly.
2. Simulation works the same way.
3. Our trained policy doesn't control hands.
4. Yes.
5. Our release only includes support for the Dex3 hand. If you would like to perform a hand swap, please contact the corresponding hand supplier for guidance and SDK support.

**Comment by Gumin-Research** [comment]
> 1. The hand action is passed directly from the interface to the robot; it is not controlled by the Sonic policy. If you are using the VLA setup, then the VLA controls it directly.
> 2. Simulation works the same way.
> 3. Our trained policy doesn't control hands.
> 4. Yes.
> 5. Our release only includes support for the Dex3 hand. If you would like to perform a hand swap, please contact the corresponding hand supplier for guidance and SDK support.

Thanks for the clear answer. While reading the decoder code I came across something I wanted to ask about.

You mentioned that the policy does not control hands. But while tracing the decoder code, I noticed two things that seem to suggest the framework has wiring for hand-aware decoders, and I wanted to ask whether these paths actually route hand through a decoder or not.

In `universal_token_modules.py:909-921`, the action assembly has a hand concat path:

```python
if "body_action" in g1_dyn_out:
    action_mean = g1_dyn_out["body_action"]
elif "meta_action" in g1_dyn_out:
    action_mean = g1_dyn_out["meta_action"]
else:
    action_mean = g1_dyn_out["action"]

if "hand_dyn" in decoded_outputs and "hand_action" in decoded_outputs["hand_dyn"]:
    hand_action = decoded_outputs["hand_dyn"]["hand_action"]
    action_mean = torch.cat([action_mean, hand_action], dim=-1)
```

And the dim defaults (line 342-343):

```python
"hand_action": hand_action_dim or 14,  # 14 hand joints default
"body_action": body_action_dim or 29,  # 29 body joints default
```

Reading just this, it looks like the framework can emit a 43-dim action (29 body + 14 hand), either via a unified `g1_dyn` with `actions_dim=43`, or via `g1_dyn` (29) plus a separate `hand_dyn` (14) concatenated. So my questions are:

1. **What is `hand_dyn`?** There is no `hand_dyn` decoder class, no `hand_dyn` config file, and no entry referencing `hand_dyn` in any released config (`all_mlp_v1.yaml`, etc.). The only references are the two `if "hand_dyn" in ...` guards at line 919 and 1158, which short-circuit when nothing populates the key. Is `hand_dyn` an empty slot that the released training simply never fills, or is there a non-released decoder definition I am missing?

2. **What is the 43-dim path for?** Given the 29 / 14 defaults and the concat code look intentional, is the 43-dim path (unified `g1_dyn` at 43, or split `g1_dyn` + `hand_dyn`) a planned extension slot for future hand-aware policies, or leftover scaffolding from earlier experiments?

3. **Does hand actually go through a decoder anywhere?** Following from above, in the released setup `g1.py` registers 29 actuated joints with no finger actuators, so `actions_dim` resolves to 29 and `g1_dyn` outputs 29-dim. Combined with `hand_dyn` not being instantiated, my reading is that the 14-dim hand never passes through any decoder in this release, matching your "Our trained policy doesn't control hands" answer. Just want to confirm this reading is correct.

Not a support request, just trying to understand the design. Thanks.
